File: api/generate.js
export default async function handler(req, res) {
    if (req.method !== 'POST') {
        res.setHeader('Allow', ['POST']);
        return res.status(405).json({ success: false, error: `Method ${req.method} Not Allowed` });
    }

    try {
        const { userQuery, base64DataOnly, mimeType, currentAspectRatio } = req.body;
        const apiKey = process.env.GEMINI_API_KEY;

        if (!apiKey) {
            return res.status(500).json({ success: false, error: 'Missing GEMINI_API_KEY environment variable on Vercel.' });
        }

        const parts = [];
        if (base64DataOnly) {
            parts.push({
                inlineData: {
                    mimeType: mimeType || 'image/jpeg',
                    data: base64DataOnly
                }
            });
        }
        parts.push({ text: userQuery || "Analyze image and generate realistic photography prompt." });

        const payload = {
            contents: [{ parts }]
        };

        const geminiUrl = `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash-preview-09-2025:generateContent?key=${apiKey}`;

        const response = await fetch(geminiUrl, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });

        const data = await response.json();

        if (data.candidates && data.candidates[0]?.content?.parts?.[0]?.text) {
            let generatedText = data.candidates[0].content.parts[0].text;
            if (currentAspectRatio && !generatedText.includes(currentAspectRatio)) {
                generatedText += ` ${currentAspectRatio}`;
            }
            return res.status(200).json({ success: true, text: generatedText });
        } else {
            console.error('Gemini API Response Error:', JSON.stringify(data));
            return res.status(500).json({ success: false, error: data.error?.message || 'Failed to generate prompt from Gemini API.' });
        }

    } catch (error) {
        console.error('Serverless Function Error:', error);
        return res.status(500).json({ success: false, error: error.message || 'Internal Server Error' });
    }
}

File: api/generate.py
import os
import json
from http.server import BaseHTTPRequestHandler
import google.generativeai as genai
from PIL import Image
import io
import base64

# កំណត់ Gemini API Key ពី Environment Variable របស់ Vercel
genai.configure(api_key=os.environ.get("GEMINI_API_KEY"))

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            # អានទិន្នន័យ JSON ដែលផ្ញើមកពី Frontend
            content_length = int(self.headers.get('Content-Length', 0))
            post_data = self.rfile.read(content_length)
            data = json.loads(post_data.decode('utf-8'))

            user_query = data.get('userQuery', '')
            base64_data = data.get('base64DataOnly', None)
            mime_type = data.get('mimeType', 'image/jpeg')

            # ជ្រើសរើសម៉ូដែល Gemini 1.5 Flash ຕາມការណែនាំ
            model = genai.GenerativeModel('gemini-1.5-flash')

            content_parts = [user_query]

            # ប្រសិនបើមានរូបភាពយោង ត្រូវបម្លែងវាដើម្បីបញ្ជូនទៅ Vision API
            if base64_data:
                image_bytes = base64.b64decode(base64_data)
                image = Image.open(io.BytesIO(image_bytes))
                content_parts.append(image)

            # ហៅទៅកាន់ Gemini API
            response = model.generate_content(content_parts)
            generated_text = response.text

            # ส่งលទ្ធផលត្រឡប់ទៅ Frontend វិញ
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            response_data = {"success": True, "text": generated_text}
            self.wfile.write(json.dumps(response_data).encode('utf-8'))

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            error_data = {"success": False, "error": str(e)}
            self.wfile.write(json.dumps(error_data).encode('utf-8'))